File: CityTaxApp/app/src/main/java/com/sgs/citytax/util/Pagination.java
package com.sgs.citytax.util;


/*
 * Created by ADIL on 20-12-2018.
 */


import android.util.Log;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class Pagination {
    private static final String TAG = "Pagination";
    private int default_initialPageNumber = 1;
    private int default_defaultPageSize = 10;

    private int initialPageNumber = 1;
    //    private int lastPageNumber = -1;
    private int defaultPageSize = 10;
    private int totalRecords = 0;
    private boolean stop_pagination;
    private boolean isScrolled = false;

    private final RecyclerView recyclerView;
    private final PaginationListener paginationListener;

    public Pagination(int initialPageNumber, int defaultPageSize, RecyclerView recyclerView, PaginationListener paginationListener) {
        this.initialPageNumber = initialPageNumber;
        this.defaultPageSize = defaultPageSize;
        this.default_initialPageNumber = initialPageNumber;
        this.default_defaultPageSize = defaultPageSize;
        this.recyclerView = recyclerView;
        this.paginationListener = paginationListener;

        setListener();
    }

    public void resetInitialPageNumber() {
        this.initialPageNumber = 1;
    }
    public int getInitialPageNumber() {
        return initialPageNumber;
    }

    public int getDefaultPageSize() {
        return defaultPageSize;
    }

    public void setDefaultValues() {
        this.initialPageNumber = default_initialPageNumber;
        this.defaultPageSize = default_defaultPageSize;
        stop_pagination = true;
        isScrolled = false;
        paginationListener.pageToCall(initialPageNumber, defaultPageSize);
    }


    private void setListener() {
        if (recyclerView == null) {
            return;
        }

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                int visibleItemCount = recyclerView.getLayoutManager().getChildCount();
                int totalItemCount = recyclerView.getLayoutManager().getItemCount();
                int firstVisibleItemPosition = 0;
                if (recyclerView.getLayoutManager() instanceof LinearLayoutManager) {
                    firstVisibleItemPosition = ((LinearLayoutManager) recyclerView.getLayoutManager()).findFirstVisibleItemPosition();
                } else if (recyclerView.getLayoutManager() instanceof GridLayoutManager) {
                    firstVisibleItemPosition = ((GridLayoutManager) recyclerView.getLayoutManager()).findFirstVisibleItemPosition();
                }

                if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount - 3 && firstVisibleItemPosition >= 0) {
                    if (!stop_pagination && !isScrolled && totalItemCount != getTotalRecords()) {
                        initialPageNumber++;
                        isScrolled = true;
                        paginationListener.pageToCall(initialPageNumber, defaultPageSize);
                    }
                }
            }
        });
    }

    public void stopPagination(int size) {
        if (size == 0) {
            stop_pagination = true;
        } else {
            stop_pagination = size % defaultPageSize != 0 || defaultPageSize == getTotalRecords();
        }
    }

    public void setIsScrolled(boolean isScrolled) {
        this.isScrolled = isScrolled;
    }

    public int getTotalRecords() {
        return totalRecords;
    }

    public void setTotalRecords(int totalRecords) {
        this.totalRecords = totalRecords;
    }

    public void doNextCall() {
        int totalItemCount = recyclerView.getLayoutManager().getItemCount();
        if (totalItemCount < getTotalRecords()) {
            initialPageNumber++;
            paginationListener.pageToCall(initialPageNumber, defaultPageSize);
        } else {
            Log.d(TAG, "nextCall: finished");
        }
    }

    public void doForceNextCall() {
        initialPageNumber++;
        paginationListener.pageToCall(initialPageNumber, defaultPageSize);
    }

    public interface PaginationListener {
        void pageToCall(int pageNumber, int PageSize);
    }

}

File: HiltArchi/app/src/main/java/com/pru/hiltarchi/utils/Pagination.kt
package com.pru.hiltarchi.utils

import android.util.Log
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class Pagination(
    initialPageNumber: Int,
    defaultPageSize: Int,
    recyclerView: RecyclerView?,
    listener: (Int, Int) -> Unit
) {
    private var defaultInitialpagenumber = 1
    private var defaultDefaultpagesize = 10
    var initialPageNumber = 1
        private set
    var defaultPageSize = 10
        private set
    var totalRecords = 0
    private var stop_pagination = false
    private var isScrolled = false
    private val recyclerView: RecyclerView?
    private val listener: (Int, Int) -> Unit
    private val TAG = "Pagination"

    init {
        this.initialPageNumber = initialPageNumber
        this.defaultPageSize = defaultPageSize
        defaultInitialpagenumber = initialPageNumber
        defaultDefaultpagesize = defaultPageSize
        this.recyclerView = recyclerView
        this.listener = listener
        setListener()
    }

    fun resetInitialPageNumber() {
        initialPageNumber = 1
    }

    fun setDefaultValues() {
        initialPageNumber = defaultInitialpagenumber
        defaultPageSize = defaultDefaultpagesize
        stop_pagination = true
        isScrolled = false
        listener.invoke(initialPageNumber, defaultPageSize)
    }

    private fun setListener() {
        if (recyclerView == null) {
            return
        }
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val visibleItemCount = recyclerView.layoutManager!!.childCount
                val totalItemCount = recyclerView.layoutManager!!.itemCount
                var firstVisibleItemPosition = 0
                if (recyclerView.layoutManager is LinearLayoutManager) {
                    firstVisibleItemPosition =
                        (recyclerView.layoutManager as LinearLayoutManager?)!!.findFirstVisibleItemPosition()
                } else if (recyclerView.layoutManager is GridLayoutManager) {
                    firstVisibleItemPosition =
                        (recyclerView.layoutManager as GridLayoutManager?)!!.findFirstVisibleItemPosition()
                }
                if (visibleItemCount + firstVisibleItemPosition >= totalItemCount - 3 && firstVisibleItemPosition >= 0) {
                    if (!stop_pagination && !isScrolled && totalItemCount != totalRecords) {
                        initialPageNumber++
                        isScrolled = true
                        listener.invoke(initialPageNumber, defaultPageSize)
                    }
                }
            }
        })
    }

    fun stopPagination(size: Int) {
        stop_pagination = if (size == 0) {
            true
        } else {
            size % defaultPageSize != 0 || defaultPageSize == totalRecords
        }
    }

    fun setIsScrolled(isScrolled: Boolean) {
        this.isScrolled = isScrolled
    }

    fun doNextCall() {
        val totalItemCount = recyclerView!!.layoutManager!!.itemCount
        if (totalItemCount < totalRecords) {
            initialPageNumber++
            listener.invoke(initialPageNumber, defaultPageSize)
        } else {
            Log.d(TAG, "nextCall: finished")
        }
    }

    fun doForceNextCall() {
        initialPageNumber++
        listener.invoke(initialPageNumber, defaultPageSize)
    }

}